kong/runloop/wasm.lua

```diff
@@ -0,0 +1,646 @@
+  end
+
+  STATE = new
+end
+
+
+---
+-- Conditionally rebuild and update the filter chain state.
+--
+-- If the current state matches the desired version, no update
+-- will be performed.
+--
+---@param  new_version? string
+---@return boolean?     ok
+---@return string?      error
+local function update_in_place(new_version)
+  if not ENABLED then
+    return true
+  end
+
+  new_version = new_version or get_version()
+  local old = STATE
+
+  if new_version == old.version then
+    log(DEBUG, "filter chain state is already up-to-date, no changes needed")
+    return true
+  end
+
+  local new, err = rebuild_state(kong.db, new_version, old)
+  if not new then
+    log(ERR, "failed rebuilding filter chain state: ", err)
+    return nil, err
+  end
+
+  set_state(new)
+
+  return true
+end
+
+
+
+---@param route?    { id: string }
+---@param service?  { id: string }
+---@return kong.runloop.wasm.filter_chain_reference?
+local function get_filter_chain_for_request(route, service)
+  local service_id = service and service.id
+  local route_id = route and route.id
+  local state = STATE
+
+  return get_chain_ref(state, TYPE_COMBINED, service_id, route_id)
+      or get_chain_ref(state, TYPE_SERVICE, service_id)
+      or get_chain_ref(state, TYPE_ROUTE, nil, route_id)
+end
+
+
+_M.get_version = get_version
+
+_M.update_in_place = update_in_place
+
+_M.set_state = set_state
+
+
+---@param kong_config table
+function _M.init(kong_config)
+  if not kong_config.wasm then
+    return
+  end
+
+  local modules = kong_config.wasm_modules_parsed
+  if not modules or #modules == 0 then
+    return
+  end
+
+  -- setup a DNS client for ngx_wasm_module
+  _G.dns_client = dns(kong_config)
+
+  proxy_wasm = require "resty.http.proxy_wasm"
+  ENABLED = true
+
+  ATTACH_OPTS.isolation = proxy_wasm.isolations.FILTER
+end
+
+
+---@return boolean? ok
+---@return string?  error
+function _M.init_worker()
+  if not ENABLED then
+    return true
+  end
+
+  local ok, err = update_in_place()
+  if not ok then
+    return nil, err
+  end
+
+  return true
+end
+
+
+local function set_proxy_wasm_property(property, value)
+  if not value then
+    return
+  end
+
+  local ok, err = proxy_wasm.set_property(property, value)
+  if not ok then
+    log(ERR, "failed to set ", property, " property: ", err)
+  end
+end
+
+
+---
+-- Lookup and execute the filter chain that applies to the current request
+-- (if any).
+--
+---@param ctx table # the request ngx.ctx table
+function _M.attach(ctx)
+  if not ENABLED then
+    return
+  end
+
+  local chain = get_filter_chain_for_request(ctx.route, ctx.service)
+
+  if not chain then
+    -- no matching chain for this route/service
+    return
+  end
+
+  if not chain.c_plan then
+    -- all filters in this chain are disabled
+    return
+  end
+
+  local ok, err = proxy_wasm.attach(chain.c_plan, ATTACH_OPTS)
+  if not ok then
+    log(ERR, "failed attaching ", chain.label, " filter chain to request: ", err)
+    error(err)
```

Comment:
So this will produce an `error` log, plus another `error` log with stacktrace but a smaller version of the error? There's also no mention of "wasm", not sure if intentional once more. There's another instance of this double-error + stacktrace below.

imho a single error, fully-formatted, perhaps higher than `err`, probably without stacktrace (unless we need it because this might have multiple callers?) would be nicer.

Reply:
Thanks. This is only called in the runloop/request context, so I _think_ a good change would just be use `kong.response.error()` to return a 500/ISE in place of throwing with `error()`--do you agree? The primary concern to me is just ensuring that the request is terminated instead of proxied.